frida_tools/apk.py

```diff
@@ -0,0 +1,313 @@
+        chunk_data[8:8 + 4] = struct.pack("<I", self.string_count)
+
+        # Increase strings offset since we have inserted a new offset and thus
+        # shifted the offset of the strings
+        self.strings_offset += 4
+        chunk_data[20:20 + 4] = struct.pack("<I", self.strings_offset)
+
+        # If there are styles, offset them as we have inserted into the strings
+        # offsets
+        if self.styles_offset != 0:
+            self.styles_offset += 4
+            chunk_data[24:24 + 4] = struct.pack("<I", self.strings_offset)
+
+        self.header.chunk_data = bytes(chunk_data)
+
+        self.dirty = True
+
+        return self.string_count - 1
+
+
+class BinaryXML:
+    def __init__(self, stream: BufferedReader):
+        self.stream = stream
+        self.chunk_headers = []
+        self.parse()
+
+    def parse(self) -> None:
+        chunk_header = ChunkHeader(self.stream, False)
+        if chunk_header.type != ChunkType.XML:
+            raise BadHeader()
+        self.chunk_headers.append(chunk_header)
+
+        size = chunk_header.size
+
+        while self.stream.tell() < size:
+            chunk_header = ChunkHeader(self.stream)
+            self.chunk_headers.append(chunk_header)
+
+
+def debug(path: str, output_path: str) -> None:
+    with ZipFile(path, "r") as iz, ZipFile(output_path, "w") as oz:
+        for info in iz.infolist():
+            with iz.open(info) as f:
+                if info.filename == "AndroidManifest.xml":
+                    manifest = BinaryXML(f)
+
+                    pool = None
+                    debuggable_index = None
+
+                    size = 8
+                    for header in manifest.chunk_headers[1:]:
+                        if header.type == ChunkType.STRING_POOL:
+                            pool = StringPool(header)
+                            debuggable_index = pool.append_str("debuggable")
+
+                        if header.type == ChunkType.START_ELEMENT:
+                            start = StartElement(header)
+                            name = pool.get_string(start.name)
+                            if name == "application":
+                                start.insert_debuggable(debuggable_index)
+
+                        size += header.size
+
+                    header = manifest.chunk_headers[0]
+                    header_data = bytearray(header.chunk_data)
+                    header_data[4:4 + 4] = struct.pack("<I", size)
+
+                    data = bytearray()
+                    data.extend(header_data)
+                    for header in manifest.chunk_headers[1:]:
+                        data.extend(header.chunk_data)
+
+                    oz.writestr(info.filename, bytes(data), info.compress_type)
+                elif info.filename.startswith("META-INF"):
+                    pass
+                else:
+                    oz.writestr(info.filename, f.read(), info.compress_type)
+
+
+def main() -> None:
```

Comment:
Would be good to move this function to the top — should be able to start reading from the top, where classes/components used by the code at the top follows below it, from highest to lowest level details. (Often referred to as [newspaper code structure](https://kentcdodds.com/blog/newspaper-code-structure)).